src/entity.ts

```diff
@@ -376,11 +435,49 @@ export namespace entity {
     return value instanceof entity.Key;
   }
 
+  /**
+   * Convert a protobuf `integerValue`.
+   *
+   * @private
+   * @param {object} value The `integerValue` to convert.
+   */
+  function decodeIntegerValue(value: ValueProto) {
+    const num = Number(value.integerValue);
+    if (!Number.isSafeInteger(num)) {
+      throw new Error(
+        'We attempted to return all of the numeric values, but ' +
+          (value.propertyName ? value.propertyName + ' ' : '') +
+          'value ' +
+          value.integerValue +
+          " is out of bounds of 'Number.MAX_SAFE_INTEGER'.\n" +
+          "Please consider passing 'options.wrapNumbersOptions=true' or\n" +
+          "'options.wrapNumbersOptions' as\n" +
+          '{\n' +
+          '  integerTypeCastFunction: provide <your_custom_function>\n' +
+          '  properties: optionally specify property name(s) to be cutom casted' +
+          '}\n' +
+          'to prevent this error.'
+      );
+    }
+    return num;
+  }
+
+  /**
+   * @typedef {object} IntegerTypeCastOptions Configuration to convert
+   *     values of `integerValue` type to a custom value. Must provide an
+   *     `integerTypeCastFunction` to handle `integerValue` conversion.
+   * @property {function} integerTypeCastFunction A custom user
```

Comment:
I think users will be surprised when they learn `integerValue`s are not converted to what they returned from their function until "valueOf()" is called. Imagining a scenario where a user has their own Int implementation, `AppBigInt`, I think they would expect `instanceof entity.data.myNumber === AppBigInt` rather than our `DatastoreInt`-- which they probably assumed they were opting out of.

If you still think having it work the way it does in this PR now would be best, let's beef up the docs to over-explain how their return value will come into play.

Reply:
Using the logic flow you wrote above, this is my proposal:

1. Default - `query.run(q)`
    - API will try to convert to `Number`
    - Will throw on overflow
2. Pass options.wrapNumbers as boolean (just wrap numbers, no custom cast) - `query.run(q, {wrapNumbers: true})`
    - API will return a DsInt obj in which
      - #valueOf will act the same way as default behavior, i.e. try to convert to Number, throw on overflow. However now, in case of an exception with #valueOf user can still retrieve a string representation of the numerical value with `#value` and does not need to make another network call.
3. Pass options.wrapNumbers as object - `query.run(q, {wrapNumbers: {integerTypeCastFunction: = myFunction}})`
    - API will return the return value from user-provided `wrapNumbers.integerTypeCastFunction`

Reply:
Implemented